File: thriving-village/src/lib/circuit-breaker.ts
/**
 * Minimal circuit breaker with a bulkhead (concurrency cap). No external
 * dependency — this app only needs three states and a failure counter.
 *
 * CLOSED -> (failureThreshold consecutive failures) -> OPEN
 * OPEN -> (cooldownMs elapsed) -> HALF_OPEN (lets exactly one trial through)
 * HALF_OPEN -> (halfOpenSuccesses consecutive successes) -> CLOSED
 * HALF_OPEN -> (any failure) -> OPEN, cooldown restarts
 *
 * The bulkhead is checked before the breaker state: a call rejected for being
 * over maxConcurrent never ran, so it isn't counted as a breaker failure.
 */

export class CircuitOpenError extends Error {
  constructor(message = 'Circuit breaker is open') {
    super(message);
    this.name = 'CircuitOpenError';
  }
}

export class BulkheadRejectedError extends Error {
  constructor(message = 'Too many concurrent calls to this dependency') {
    super(message);
    this.name = 'BulkheadRejectedError';
  }
}

type BreakerOptions = {
  /** Consecutive failures while CLOSED before tripping to OPEN. */
  failureThreshold: number;
  /** How long OPEN lasts before a single HALF_OPEN trial is allowed. */
  cooldownMs: number;
  /** Consecutive HALF_OPEN successes required to fully CLOSE. */
  halfOpenSuccesses: number;
  /** In-flight calls beyond this are rejected immediately, not queued. */
  maxConcurrent: number;
};

type State = 'closed' | 'open' | 'half-open';

export class CircuitBreaker {
  private state: State = 'closed';
  private consecutiveFailures = 0;
  private consecutiveSuccesses = 0;
  private openedAt = 0;
  private inFlight = 0;
  private halfOpenTrialInFlight = false;

  constructor(private readonly opts: BreakerOptions) {}

  async exec<T>(fn: () => Promise<T>): Promise<T> {
    if (this.inFlight >= this.opts.maxConcurrent) {
      throw new BulkheadRejectedError();
    }

    if (this.state === 'open') {
      if (Date.now() - this.openedAt < this.opts.cooldownMs) {
        throw new CircuitOpenError();
      }
      this.state = 'half-open';
      this.consecutiveSuccesses = 0;
    }

    if (this.state === 'half-open') {
      // Only one trial call at a time while probing recovery.
      if (this.halfOpenTrialInFlight) throw new CircuitOpenError();
      this.halfOpenTrialInFlight = true;
    }

    this.inFlight++;
    try {
      const result = await fn();
      this.onSuccess();
      return result;
    } catch (err) {
      this.onFailure();
      throw err;
    } finally {
      this.inFlight--;
      if (this.state === 'half-open') this.halfOpenTrialInFlight = false;
    }
  }

  private onSuccess() {
    if (this.state === 'half-open') {
      this.consecutiveSuccesses++;
      if (this.consecutiveSuccesses >= this.opts.halfOpenSuccesses) {
        this.state = 'closed';
        this.consecutiveFailures = 0;
        this.consecutiveSuccesses = 0;
      }
      return;
    }
    this.consecutiveFailures = 0;
  }

  private onFailure() {
    if (this.state === 'half-open') {
      this.state = 'open';
      this.openedAt = Date.now();
      return;
    }
    this.consecutiveFailures++;
    if (this.consecutiveFailures >= this.opts.failureThreshold) {
      this.state = 'open';
      this.openedAt = Date.now();
    }
  }
}

File: thriving-village/backend_django/apps/integrations/circuit_breaker.py
"""Line-for-line port of backend/src/utils/circuit-breaker.ts.

CLOSED -> (failureThreshold consecutive failures) -> OPEN
OPEN -> (cooldown elapsed) -> HALF_OPEN (lets exactly one trial through)
HALF_OPEN -> (halfOpenSuccesses consecutive successes) -> CLOSED
HALF_OPEN -> (any failure) -> OPEN, cooldown restarts

The bulkhead (maxConcurrent) is checked before the breaker state: a call
rejected for being over the concurrency cap never ran, so it isn't counted
as a breaker failure. A lock guards the counters since Django may run
multiple threads per worker (Strapi's original only needed this for
single-threaded Node).
"""

import threading
import time


class CircuitOpenError(Exception):
    pass


class BulkheadRejectedError(Exception):
    pass


class CircuitBreaker:
    def __init__(self, failure_threshold: int, cooldown_seconds: float, half_open_successes: int, max_concurrent: int):
        self.failure_threshold = failure_threshold
        self.cooldown_seconds = cooldown_seconds
        self.half_open_successes = half_open_successes
        self.max_concurrent = max_concurrent

        self._lock = threading.Lock()
        self._state = "closed"
        self._consecutive_failures = 0
        self._consecutive_successes = 0
        self._opened_at = 0.0
        self._in_flight = 0
        self._half_open_trial_in_flight = False

    def exec(self, fn):
        with self._lock:
            if self._in_flight >= self.max_concurrent:
                raise BulkheadRejectedError()

            if self._state == "open":
                if time.monotonic() - self._opened_at < self.cooldown_seconds:
                    raise CircuitOpenError()
                self._state = "half-open"
                self._consecutive_successes = 0

            if self._state == "half-open":
                if self._half_open_trial_in_flight:
                    raise CircuitOpenError()
                self._half_open_trial_in_flight = True

            self._in_flight += 1

        try:
            result = fn()
            self._on_success()
            return result
        except Exception:
            self._on_failure()
            raise
        finally:
            with self._lock:
                self._in_flight -= 1
                if self._state == "half-open":
                    self._half_open_trial_in_flight = False

    def _on_success(self):
        with self._lock:
            if self._state == "half-open":
                self._consecutive_successes += 1
                if self._consecutive_successes >= self.half_open_successes:
                    self._state = "closed"
                    self._consecutive_failures = 0
                    self._consecutive_successes = 0
                return
            self._consecutive_failures = 0

    def _on_failure(self):
        with self._lock:
            if self._state == "half-open":
                self._state = "open"
                self._opened_at = time.monotonic()
                return
            self._consecutive_failures += 1
            if self._consecutive_failures >= self.failure_threshold:
                self._state = "open"
                self._opened_at = time.monotonic()
